### DISCUSSION TITLE
help wanted: Pipeline custom provider

### DISCUSSION THREAD

**Discussion by Shayano** [open-webui/open-webui#3463] [Q&A] [answered]
#### Problem Description

Hello, I am trying to set up a custom pipeline for the Infomaniak provider using Open WebUI. The process works with the "Provider v1" example provided at the end of this document. However, when the context or message content is too large, I encounter the following error:

```
Error: 400 Client Error: Bad Request for url: https://api.infomaniak.com/1/ai/100083/openai/chat/completions
```

I have also tried using a different variant with "Provider v2" where I set the maximum tokens to 5000, thinking that the issue might be related to token limits. The problem changed, and now I receive the error:

```
Unexpected token '<', "<!DOCTYPE "... is not valid JSON
```

Shorter messages still work fine in both cases.

I have also attempted to add the provider to AnythingLLM, and it works without any modifications using the same messages. However, I have been struggling with this issue for days trying to make it work via an Open WebUI pipeline.

##### Provider v1

```python
from typing import List, Union, Generator, Iterator
from schemas import OpenAIChatMessage
from pydantic import BaseModel
import os
import requests

class Pipeline:
    class Valves(BaseModel):
        INFOMANIAK_API_KEY: str = ""
        PRODUCT_ID: int = 0
        MODEL: str = ""

    def __init__(self):
        self.name = "Infomaniak"
        self.valves = self.Valves(
            **{
                "INFOMANIAK_API_KEY": os.getenv("INFOMANIAK_API_KEY", "infomaniak api key here"),
                "PRODUCT_ID": int(os.getenv("PRODUCT_ID", 0)),
                "MODEL": os.getenv("MODEL", "mixtral, mixtral8x22b or llama3"),
            }
        )
        pass

    async def on_startup(self):
        print(f"on_startup:{__name__}")
        pass

    async def on_shutdown(self):
        print(f"on_shutdown:{__name__}")
        pass

    def pipe(
        self, user_message: str, model_id: str, messages: List[dict], body: dict
    ) -> Union[str, Generator, Iterator]:
        print(f"pipe:{__name__}")

        print(messages)
        print(user_message)

        INFOMANIAK_API_KEY = self.valves.INFOMANIAK_API_KEY
        MODEL = self.valves.MODEL
        PRODUCT_ID = self.valves.PRODUCT_ID

        headers = {}
        headers["Authorization"] = f"Bearer {INFOMANIAK_API_KEY}"
        headers["Content-Type"] = "application/json"

        payload = {
            **body,
            "model": MODEL,
            "messages": messages
        }

        if "user" in payload:
            del payload["user"]
        if "chat_id" in payload:
            del payload["chat_id"]
        if "title" in payload:
            del payload["title"]

        print(payload)

        try:
            r = requests.post(
                url=f"https://api.infomaniak.com/1/ai/{PRODUCT_ID}/openai/chat/completions",
                json=payload,
                headers=headers,
                stream=True,
            )

            r.raise_for_status()

            if body["stream"]:
                return r.iter_lines()
            else:
                return r.json()
        except Exception as e:
            return f"Error: {e}"
```

##### Provider v2

```python
from typing import List, Union, Generator, Iterator
from schemas import OpenAIChatMessage
from pydantic import BaseModel
import os
import requests

class Pipeline:
    class Valves(BaseModel):
        INFOMANIAK_API_KEY: str = ""
        PRODUCT_ID: int = 0
        MODEL: str = "mixtral"

    def __init__(self):
        self.name = "Infomaniak new"
        self.valves = self.Valves(
            **{
                "INFOMANIAK_API_KEY": os.getenv("INFOMANIAK_API_KEY", "infomaniak-api-key-here"),
                "PRODUCT_ID": int(os.getenv("PRODUCT_ID", 0)),
                "MODEL": os.getenv("MODEL", "mixtral"),
            }
        )
        pass

    async def on_startup(self):
        print(f"on_startup:{__name__}")
        pass

    async def on_shutdown(self):
        print(f"on_shutdown:{__name__}")
        pass

    def pipe(self, user_message: str, model_id: str, messages: List[dict], body: dict) -> Union[str, Generator, Iterator]:
        print(f"pipe:{__name__}")

        print("Messages received:", messages)
        print("User message:", user_message)

        INFOMANIAK_API_KEY = self.valves.INFOMANIAK_API_KEY
        MODEL = self.valves.MODEL
        PRODUCT_ID = self.valves.PRODUCT_ID

        headers = {
            "Authorization": f"Bearer {INFOMANIAK_API_KEY}",
            "Content-Type": "application/json"
        }

        payload = {
            "model": MODEL,
            "messages": messages,
            "max_tokens": body.get("max_tokens", 5000),
            "temperature": body.get("temperature", 0.7),
            "stream": body.get("stream", True),
        }

        # Remove unnecessary fields
        payload.pop("user", None)
        payload.pop("chat_id", None)
        payload.pop("title", None)

        if 'profile_type' in body:
            payload['profile_type'] = body['profile_type']

        print("Payload sent:", payload)

        try:
            r = requests.post(
                url=f"https://api.infomaniak.com/1/ai/{PRODUCT_ID}/openai/chat/completions",
                json=payload,
                headers=headers,
                stream=True,
            )

            r.raise_for_status()

            if body["stream"]:
                return r.iter_lines()
            else:
                return r.json()

        except Exception as e:
            return f"Error: {e}"
```

If anyone could provide guidance, I would greatly appreciate it. I have been struggling with this issue for days, and while it works perfectly on AnythingLLM with the same messages, I can't seem to get it working via an Open WebUI pipeline.
Thanks for your help.
---

If there are inconsistencies in the code, it's because I'm not a developer and I'm tinkering with AI to get what I want.

**Comment by justinh-rahb** [comment]
If these are just OpenAI compatible API endpoints then why do you need a custom pipe at all? Does it not work if you put the appropriate base URL and key into **Settings > OpenAI connections**? Unless I'm missing something else here..

**Comment by Shayano** [comment]
Thanks for your answer, 
Unfortunately it doesn't work.

Maybe I'm missing something, but to retrieve the list of models you need to use the url:
https://api.infomaniak.com/1/ai

But for chat you need to add a product ID which is specific to each user using their service:
https://api.infomaniak.com/1/ai/{PRODUCT_ID}/openai/chat/completions

That's why I decided to use a pipeline, but maybe there's a simpler solution and I'd have missed it.

doc ref:
https://developer.infomaniak.com/docs/api/get/1/ai
https://developer.infomaniak.com/docs/api/post/1/ai/%7Bproduct_id%7D/openai/chat/completions

**Reply by justinh-rahb** [reply]
Ahh indeed that's what I missed. Yep, this would require a pipeline. Likely would be able to just modify the existing OpenAI one with the custom paths.

The issue you're having with your current snippets above seems to be in error handling, you seem to be getting responses which would indicate that it crashed. The full logs from Pipelines and WebUI containers would be most helpful in resolving the exact cause.

**Comment by Shayano** [comment]
Thanks for pointing me in the right direction, tomorrow I'm trying to get logs for the different scenarios but I've already been able to get a more explicit error.

`Response content: b'{"result":"error","error":{"code":"bad_request","description":"Your inputs tokens + max_tokens = 320790 but must be < 23000"}}'`

I had already tried to split the content but without success for the moment. If anyone knows how to do it, I'd love to hear from you.

**Comment by Shayano** [comment, marked as answer]
If anyone ever needed it, all they had to do was modify the maximum context according to the model used:
This is the "Context Length" parameter (num_ctx) in the model's advanced options.
mixtral8x22b = 23k
mixtral = 32k
llama3 = 8k
